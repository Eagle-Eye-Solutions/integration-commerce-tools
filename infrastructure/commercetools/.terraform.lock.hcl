# This file is maintained automatically by "terraform init".
# Manual edits may be lost in future updates.

provider "registry.terraform.io/labd/commercetools" {
  version = "1.12.0"
  hashes = [
    "h1:YWpFdVmqiWaUNBphlqOjoboKIK6c4KVMKQDQLtNnaPE=",
    "zh:01361b3324953d85c3c0640acc47e6c9c982c422f7d4b216bf3d333c967ce97f",
    "zh:196d44ce9eb65b34dc6ff423fd5a4394e837384a1776ad9a7cde1e5778d53c76",
    "zh:3c6cd2dec4d3e8ee3becc3ac8c3d8e452595fb06fcb1b5e9f1e9bf1db204f270",
    "zh:4a4bfb39c584a5c349aa5bc60e5530b02ffa8fc3c2d91dfba72e89283cd5023b",
    "zh:4ce66411a5bef6fc7945ff1ea126166e9b3e40e0e023c07a6ad9fceeb700639e",
    "zh:4e579e16e608b614cc66c90834fd478058ecd593abaa3145e90893d7534273c0",
    "zh:660cd42846b03920a21a4e174438e07b045db6b421753a4b8a3569745fc1d4c9",
    "zh:7208dab6e1062b7b0160cf2cc05f43048c9db1d4e152ec8302442e7d7c91a64d",
    "zh:753f0d6bf5cf2f0393909d50886fb353efa8a3760929a1fd64e45594c3693e67",
    "zh:7bf53fc46fd46c739bdf9b7fbbbde5b39cdcd35422d2593e85d6667095a3c4f5",
    "zh:bb98a96677f81c20f1c9c7fb1a1499c1d682f7df2045bf3b3ec8e59b6edb1e31",
    "zh:c0c75d09b256b5bd2dc2081204038e72d58eaeab9827588f74a710f5485b346a",
    "zh:cb67f4323e2fa9225b5d8e88c5ecfd42dfa691a55961792fd5332aa101eb084e",
    "zh:cca2f4818baba2f4e9f45334b3c333d802e8aa640b004dd3470fb7d3cc151414",
  ]
}
