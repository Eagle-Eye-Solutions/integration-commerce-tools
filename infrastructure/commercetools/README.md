# Commercetools infrastructure

Infrastructure code to configure commercetools product types, taxes, settings, shipping methods...

### Terraform cloud setup

**NOTE:** These steps should be done only once.

* Go to [app.terraform.io](https://app.terraform.io/)
* Create a project called `eagleeye-commercetools-plugin`
* Create two workspaces called `ee-ct-plugin-commercetools-dev`, `ee-ct-plugin-commercetools-prod`
* Configure the variables required
    * CTP_PROJECT_NAME
    * CTP_CLIENT_ID
    * CTP_CLIENT_SECRET
    * CTP_PROJECT_KEY
    *
  CTP_SCOPES: `manage_project_settings:eagleeye-connector-* manage_subscriptions:eagleeye-connector-* manage_extensions:eagleeye-connector-* manage_products:eagleeye-connector-* manage_api_clients:eagleeye-connector-* manage_orders:eagleeye-connector-* manage_shipping_methods:eagleeye-connector-*`
* Connect both workspaces to the GitHub repository, in terraform cloud open the workspace,
  click `settings` > `Version Control`, connect to GitHub and set the field `Terraform Working Directory`
  to `infrastructure/commercetools`

### Login to terraform cloud

Log in to Terraform Cloud via the CLI, run `terraform login` and follow the prompts to get an API token for
Terraform to use. If you don't have a Terraform Cloud account, you can create one during this step. More info
here https://developer.hashicorp.com/terraform/tutorials/cloud-get-started/cloud-login

### Using terraform

View available workspaces: `terraform workspace list`

Select workspace: `terraform workspace select ee-ct-plugin-commercetools-dev`

View plan `terraform plan`

Apply changes `terraform apply` 
