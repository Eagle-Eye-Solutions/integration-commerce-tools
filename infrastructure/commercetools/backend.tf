# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: MPL-2.0

# The block below configures Terraform to use the 'remote' backend with Terraform Cloud.
# For more information, see https://www.terraform.io/docs/backends/types/remote.html
terraform {
  cloud {
    organization = "ApplyDigitalForEagleEye"

    workspaces {
      #      project = "eagleeye-commercetools-plugin"
      tags = ["commercetools"]
      #      prefix = "ee-ct-plugin-commercetools-"
    }
  }

  required_providers {
    commercetools = {
      source = "labd/commercetools"
    }
  }

  required_version = ">= 1.1.2"
}