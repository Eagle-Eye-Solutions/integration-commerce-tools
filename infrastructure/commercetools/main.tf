resource "commercetools_tax_category" "my-tax-category" {
  name        = "standard"
  description = "Standard Tax Category"
  key         = "standard"
}

resource "commercetools_tax_category_rate" "standard-tax-category-UK" {
  tax_category_id   = commercetools_tax_category.my-tax-category.id
  name              = "20% VAT"
  amount            = 0.20
  included_in_price = true
  country           = "GB"
}

resource "commercetools_project_settings" "my-project" {
  name                         = var.CPT_PROJECT_NAME
  countries                    = ["GB"]
  currencies                   = ["GBP"]
  languages                    = ["en"]
  enable_search_index_products = true
}

resource "commercetools_product_type" "generic-product-type" {
  key         = "generic"
  name        = "generic"
  description = "Generic product type to create test products"
}

resource "commercetools_shipping_method" "standard" {
  key            = "standard-key"
  name           = "Standard delivery"
  description    = "Standard delivery"
  localized_name = {
    en = "Standard delivery"
  }
  localized_description = {
    en = "Standard delivery"
  }
  is_default      = true
  tax_category_id = commercetools_tax_category.my-tax-category.id
  predicate       = "1 = 1"
}

resource "commercetools_shipping_zone" "gb" {
  key         = "gb"
  name        = "GB"
  description = "United Kingdom"
  location {
    country = "GB"
  }
}

resource "commercetools_shipping_zone_rate" "my-shipping-zone-rate" {
  shipping_method_id = commercetools_shipping_method.standard.id
  shipping_zone_id   = commercetools_shipping_zone.gb.id

  price {
    cent_amount   = 500
    currency_code = "GBP"
  }

  free_above {
    cent_amount   = 10000
    currency_code = "GBP"
  }
}