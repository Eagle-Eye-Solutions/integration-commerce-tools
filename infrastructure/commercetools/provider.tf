provider "commercetools" {
  api_url       = var.CTP_API_URL
  token_url     = var.CTP_AUTH_URL
  client_id     = var.CTP_CLIENT_ID
  client_secret = var.CTP_CLIENT_SECRET
  project_key   = var.CTP_PROJECT_KEY
  scopes        = var.CTP_SCOPES
}

variable "CTP_API_URL" {
  type = string
  default = "https://api.europe-west1.gcp.commercetools.com"
  description = "commercetools API URL"
}

variable "CTP_AUTH_URL" {
  type = string
  default = "https://auth.europe-west1.gcp.commercetools.com"
  description = "commercetools Authentication API URL"
}

variable "CTP_CLIENT_ID" {
  type = string
  description = "commercetools API Client ID"
  sensitive = true
}

variable "CTP_CLIENT_SECRET" {
  type = string
  description = "commercetools API Client secret"
  sensitive = true
}

variable "CTP_PROJECT_KEY" {
  type = string
  description = "commercetools API Client ID"
}

variable "CTP_SCOPES" {
  type = string
  description = "commercetools Client scopes"
}

variable "CPT_PROJECT_NAME" {
  type = string
  description = "commercetools environment project name"
}
