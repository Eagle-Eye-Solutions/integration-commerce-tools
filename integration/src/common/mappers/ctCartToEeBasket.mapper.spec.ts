import { Test, TestingModule } from '@nestjs/testing';
import { CTCartToEEBasketMapper } from './ctCartToEeBasket.mapper';
import { Commercetools } from '../../providers/commercetools/commercetools.provider';
import { ConfigService } from '@nestjs/config';
import { BASKET_STORE_SERVICE } from '../../services/basket-store/basket-store.provider';
import { BasketStoreService } from '../../services/basket-store/basket-store.interface';

describe('CTCartToEEBasketMapper', () => {
  let service: CTCartToEEBasketMapper;
  let configService: ConfigService;
  let commercetools: Commercetools;
  let basketStoreService: jest.Mocked<BasketStoreService>;

  const cart: any = {
    lineItems: [
      {
        name: {
          en: 'Example Product',
        },
        variant: {
          sku: 'SKU123',
        },
        price: {
          value: {
            centAmount: 300,
            currencyCode: 'USD',
            type: 'centPrecision',
            fractionDigits: 2,
          },
        },
        quantity: 1,
        totalPrice: {
          centAmount: 300,
          currencyCode: 'USD',
          type: 'centPrecision',
          fractionDigits: 2,
        },
      },
    ],
    totalPrice: {
      centAmount: 300,
      currencyCode: 'USD',
      type: 'centPrecision',
      fractionDigits: 2,
    },
  };
  const shippingMethodMapMock = [{ key: 'standard-key', upc: '245879' }];

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        CTCartToEEBasketMapper,
        {
          provide: Commercetools,
          useValue: {
            getShippingMethods: jest.fn(),
            getOrderById: jest.fn(),
          },
        },
        {
          provide: ConfigService,
          useValue: {
            get: jest.fn(),
          },
        },
        {
          provide: BASKET_STORE_SERVICE,
          useValue: {
            save: jest.fn(),
            get: jest.fn(),
            delete: jest.fn(),
            isEnabled: jest.fn(),
          },
        },
      ],
    }).compile();

    service = module.get<CTCartToEEBasketMapper>(CTCartToEEBasketMapper);
    configService = module.get<ConfigService>(ConfigService);
    commercetools = module.get<Commercetools>(Commercetools);
    basketStoreService = module.get(BASKET_STORE_SERVICE);
    jest.spyOn(configService, 'get').mockReturnValueOnce(shippingMethodMapMock);
    jest.resetAllMocks();
  });

  describe('mapCartLineItemsToBasketContent', () => {
    it('should return the mapped line items', () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(false)
        .mockReturnValueOnce(shippingMethodMapMock);

      const lineItems = [
        {
          variant: {
            sku: 'SKU123',
          },
          price: {
            value: {
              centAmount: 1000,
            },
          },
          totalPrice: {
            centAmount: 900,
          },
          name: {
            en: 'Product 1',
          },
          quantity: 2,
        },
      ];

      const basketContents = service.mapCartLineItemsToBasketContent(
        lineItems as any,
      );

      expect(basketContents).toMatchSnapshot();
    });
  });

  describe('mapShippingMethodSkusToBasketItems', () => {
    it('should return the mapped custom basket items', async () => {
      const shippingInfo = {
        shippingMethod: {
          id: 'some-id',
        },
        price: {
          centAmount: 300,
          currencyCode: 'USD',
          type: 'centPrecision',
          fractionDigits: 2,
        },
        shippingMethodName: 'Example Shipping Discount',
      };

      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(shippingMethodMapMock);
      jest
        .spyOn(commercetools, 'getShippingMethods')
        .mockResolvedValueOnce([{ key: 'standard-key' }] as any);

      const basketContents = await service.mapShippingMethodSkusToBasketItems(
        shippingInfo as any,
      );

      expect(basketContents).toMatchSnapshot();
    });
  });

  describe('mapAdjustedBasketToCartDirectDiscounts', () => {
    it('should return the direct discount draft', () => {
      const basket = {
        summary: {
          totalDiscountAmount: {
            promotions: 500,
          },
          adjustmentResults: [],
        },
      };

      const directDiscount = service.mapAdjustedBasketToCartDirectDiscounts(
        basket,
        cart,
      );

      expect(directDiscount).toMatchSnapshot();
    });
  });

  describe('mapAdjustedBasketToItemDirectDiscounts', () => {
    it('should return the direct discount drafts', () => {
      const basket = {
        summary: {
          totalDiscountAmount: {
            promotions: 10,
          },
        },
        contents: [
          {
            upc: 'SKU123',
            adjustmentResults: [
              {
                totalDiscountAmount: 10,
              },
            ],
          },
        ],
      };

      const directDiscounts = service.mapAdjustedBasketToItemDirectDiscounts(
        basket,
        cart,
      );

      expect(directDiscounts).toMatchSnapshot();
    });

    it('should not return the direct discount drafts if there are no adjustmentResults', () => {
      const basket = {
        summary: {
          totalDiscountAmount: {
            promotions: 10,
          },
        },
        contents: [
          {
            upc: 'SKU123',
          },
        ],
      };

      const directDiscounts = service.mapAdjustedBasketToItemDirectDiscounts(
        basket,
        cart,
      );

      expect(directDiscounts).toMatchSnapshot();
    });
  });

  describe('mapAdjustedBasketToShippingDirectDiscounts', () => {
    it('should return the direct discount drafts', () => {
      const basket = {
        summary: {
          totalDiscountAmount: {
            promotions: 10,
          },
        },
        contents: [
          {
            upc: '245879',
            adjustmentResults: [
              {
                totalDiscountAmount: 10,
              },
            ],
          },
        ],
      };

      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(shippingMethodMapMock);

      const directDiscounts =
        service.mapAdjustedBasketToShippingDirectDiscounts(basket, cart);

      expect(directDiscounts).toMatchSnapshot();
    });

    it('should not return the direct discount drafts if there are no adjustmentResults', () => {
      const basket = {
        summary: {
          totalDiscountAmount: {
            promotions: 10,
          },
        },
        contents: [
          {
            upc: '245879',
          },
        ],
      };

      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(shippingMethodMapMock);

      const directDiscounts =
        service.mapAdjustedBasketToShippingDirectDiscounts(basket, cart);

      expect(directDiscounts).toMatchSnapshot();
    });
  });

  describe('mapVoucherCodesToCampaignTokens', () => {
    it('should return an array of tokens to be examined', () => {
      const voucherCodes = ['12345678'];
      const payload = service.mapVoucherCodesToCampaignTokens(voucherCodes);

      expect(payload).toMatchSnapshot();
    });

    it('should ignore empty string vouchers', () => {
      const voucherCodes = [''];
      const payload = service.mapVoucherCodesToCampaignTokens(voucherCodes);

      expect(payload).toMatchSnapshot();
    });
  });

  describe('mapCartToWalletOpenPayload', () => {
    it('should return the payload for /wallet/open', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(false)
        .mockReturnValueOnce(shippingMethodMapMock)
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      const payload = await service.mapCartToWalletOpenPayload(cart, false);

      expect(payload).toMatchSnapshot();
    });
  });

  describe('mapOrderToWalletSettlePayload', () => {
    it('should return the payload for /wallet/settle', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      jest.spyOn(basketStoreService, 'get').mockResolvedValueOnce({
        enrichedBasket: {
          contents: [],
        },
      });

      const payload = await service.mapOrderToWalletSettlePayload({
        id: '123456',
        cart: {
          typeId: 'cart',
          id: '12345678',
        },
        custom: {
          fields: {},
        },
      } as any);

      expect(payload).toMatchSnapshot();
    });
  });

  describe('mapOrderToWalletSettlePayload', () => {
    it('should include eagleeye-identityValue for /wallet/settle when available', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      jest.spyOn(basketStoreService, 'get').mockResolvedValueOnce({
        enrichedBasket: {
          contents: [],
        },
      });

      const payload = await service.mapOrderToWalletSettlePayload({
        id: '123456',
        cart: {
          typeId: 'cart',
          id: '12345678',
        },
        custom: {
          fields: {
            'eagleeye-identityValue': 'some-identity',
          },
        },
      } as any);

      expect(payload).toMatchSnapshot();
    });
  });

  describe('mapCartToWalletOpenPayload', () => {
    it('should include voucher codes (tokens) if present in the cart', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(false)
        .mockReturnValueOnce(shippingMethodMapMock)
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      const cartWithCodes = {
        ...cart,
        custom: {
          type: {
            typeId: 'type',
            id: '123456',
          },
          fields: {
            'eagleeye-voucherCodes': ['12345678'],
          },
        },
      };

      const payload = await service.mapCartToWalletOpenPayload(
        cartWithCodes,
        false,
      );

      expect(payload).toMatchSnapshot();
    });

    it('should include potential voucher codes (old invalid tokens) if present in the cart', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(false)
        .mockReturnValueOnce(shippingMethodMapMock)
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      const cartWithCodes = {
        ...cart,
        custom: {
          type: {
            typeId: 'type',
            id: '123456',
          },
          fields: {
            'eagleeye-potentialVoucherCodes': ['1234567890'],
          },
        },
      };

      const payload = await service.mapCartToWalletOpenPayload(
        cartWithCodes,
        false,
      );

      expect(payload).toMatchSnapshot();
    });

    it('should not send to EE duplicated voucher codes when potentialVoucherCodes and voucherCodes contain the same voucher code', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(false)
        .mockReturnValueOnce(shippingMethodMapMock)
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      const cartWithCodes = {
        ...cart,
        custom: {
          type: {
            typeId: 'type',
            id: '123456',
          },
          fields: {
            'eagleeye-voucherCodes': ['1234567890'],
            'eagleeye-potentialVoucherCodes': ['1234567890'],
          },
        },
      };

      const payload = await service.mapCartToWalletOpenPayload(
        cartWithCodes,
        false,
      );

      expect(payload).toMatchSnapshot();
    });

    it('should include eagle eye identity if present in the cart', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      const cartWithCodes = {
        ...cart,
        custom: {
          type: {
            typeId: 'type',
            id: '123456',
          },
          fields: {
            'eagleeye-identityValue': ['12345678'],
          },
        },
      };

      const payload = await service.mapCartToWalletOpenPayload(
        cartWithCodes,
        true,
      );

      expect(payload).toMatchSnapshot();
    });

    it('should not include eagle eye identity if not present in the cart but the instruction is to include identity', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce('outlet1')
        .mockReturnValueOnce('banner1');

      const cartWithCodes = {
        ...cart,
        custom: {
          type: {
            typeId: 'type',
            id: '123456',
          },
          fields: {},
        },
      };

      const payload = await service.mapCartToWalletOpenPayload(
        cartWithCodes,
        true,
      );

      expect(payload).toMatchSnapshot();
    });

    it('should return the payload for /wallet/open, without the optional parentIncomingIdentifier when not set in the configuration', async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(false)
        .mockReturnValueOnce(shippingMethodMapMock)
        .mockReturnValueOnce('outlet1');

      const payload = await service.mapCartToWalletOpenPayload(cart, false);

      expect(payload).toMatchSnapshot();
    });

    it("should return the payload for /wallet/open, with sku instead of upc when 'useItemSku' is set to true", async () => {
      jest
        .spyOn(configService, 'get')
        .mockReturnValueOnce(true)
        .mockReturnValueOnce(shippingMethodMapMock)
        .mockReturnValueOnce('outlet1');

      const payload = await service.mapCartToWalletOpenPayload(cart, false);

      expect(payload).toMatchSnapshot();
    });
  });

  describe('mapAdjustedBasketToBasketEarn', () => {
    it('should return the mapped basket earn', () => {
      const basket = {
        summary: {
          adjudicationResults: [
            {
              resourceType: 'SCHEME',
              resourceId: '1653843',
              instanceId: '1653843-1',
              success: null,
              type: 'earn',
              value: null,
              balances: {
                current: 400,
              },
            },
          ],
        },
      };

      const basketContents = service.mapAdjustedBasketToBasketEarn(basket);

      expect(basketContents).toMatchSnapshot();
    });
  });
});
