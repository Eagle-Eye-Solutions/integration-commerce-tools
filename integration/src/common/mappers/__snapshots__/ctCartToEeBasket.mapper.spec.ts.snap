// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTCartToEEBasketMapper mapAdjustedBasketToBasketEarn should return the mapped basket earn 1`] = `
{
  "balance": 400,
  "offers": [],
}
`;

exports[`CTCartToEEBasketMapper mapAdjustedBasketToCartDirectDiscounts should return the direct discount draft 1`] = `[]`;

exports[`CTCartToEEBasketMapper mapAdjustedBasketToItemDirectDiscounts should not return the direct discount drafts if there are no adjustmentResults 1`] = `[]`;

exports[`CTCartToEEBasketMapper mapAdjustedBasketToItemDirectDiscounts should return the direct discount drafts 1`] = `
[
  {
    "target": {
      "predicate": "sku="SKU123"",
      "type": "lineItems",
    },
    "value": {
      "money": [
        {
          "centAmount": 10,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      ],
      "type": "absolute",
    },
  },
]
`;

exports[`CTCartToEEBasketMapper mapAdjustedBasketToShippingDirectDiscounts should not return the direct discount drafts if there are no adjustmentResults 1`] = `[]`;

exports[`CTCartToEEBasketMapper mapAdjustedBasketToShippingDirectDiscounts should return the direct discount drafts 1`] = `
[
  {
    "target": {
      "type": "shipping",
    },
    "value": {
      "money": [
        {
          "centAmount": 10,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
      ],
      "type": "absolute",
    },
  },
]
`;

exports[`CTCartToEEBasketMapper mapCartLineItemsToBasketContent should return the mapped line items 1`] = `
[
  {
    "description": "Product 1",
    "itemUnitCost": 1000,
    "itemUnitCount": 2,
    "itemUnitMetric": "EACH",
    "salesKey": "SALE",
    "totalUnitCost": 2000,
    "totalUnitCostAfterDiscount": 2000,
    "upc": "SKU123",
  },
]
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should include eagle eye identity if present in the cart 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "sku": "SKU123",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": undefined,
  "identity": {
    "identityValue": [
      "12345678",
    ],
  },
  "location": {
    "incomingIdentifier": undefined,
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should include potential voucher codes (old invalid tokens) if present in the cart 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
        "upc": "SKU123",
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": [
    {
      "type": "TOKEN",
      "value": "1234567890",
    },
  ],
  "location": {
    "incomingIdentifier": "outlet1",
    "parentIncomingIdentifier": "banner1",
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should include voucher codes (tokens) if present in the cart 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
        "upc": "SKU123",
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": [
    {
      "type": "TOKEN",
      "value": "12345678",
    },
  ],
  "location": {
    "incomingIdentifier": "outlet1",
    "parentIncomingIdentifier": "banner1",
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should not include eagle eye identity if not present in the cart but the instruction is to include identity 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "sku": "SKU123",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": undefined,
  "location": {
    "incomingIdentifier": undefined,
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should not send to EE duplicated voucher codes when potentialVoucherCodes and voucherCodes contain the same voucher code 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
        "upc": "SKU123",
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": [
    {
      "type": "TOKEN",
      "value": "1234567890",
    },
  ],
  "location": {
    "incomingIdentifier": "outlet1",
    "parentIncomingIdentifier": "banner1",
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should return the payload for /wallet/open 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
        "upc": "SKU123",
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": undefined,
  "location": {
    "incomingIdentifier": "outlet1",
    "parentIncomingIdentifier": "banner1",
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should return the payload for /wallet/open, with sku instead of upc when 'useItemSku' is set to true 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "sku": "SKU123",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": undefined,
  "location": {
    "incomingIdentifier": "outlet1",
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapCartToWalletOpenPayload should return the payload for /wallet/open, without the optional parentIncomingIdentifier when not set in the configuration 1`] = `
{
  "basket": {
    "contents": [
      {
        "description": "Example Product",
        "itemUnitCost": 300,
        "itemUnitCount": 1,
        "itemUnitMetric": "EACH",
        "salesKey": "SALE",
        "totalUnitCost": 300,
        "totalUnitCostAfterDiscount": 300,
        "upc": "SKU123",
      },
    ],
    "summary": {
      "redemptionChannel": "Online",
      "totalBasketValue": 300,
      "totalDiscountAmount": {
        "general": null,
        "promotions": 0,
        "staff": null,
      },
      "totalItems": 1,
    },
    "type": "STANDARD",
  },
  "examine": undefined,
  "location": {
    "incomingIdentifier": "outlet1",
  },
  "lock": true,
  "options": {
    "adjustBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
    "analyseBasket": {
      "enabled": true,
      "includeOpenOffers": true,
    },
  },
  "reference": undefined,
}
`;

exports[`CTCartToEEBasketMapper mapOrderToWalletSettlePayload should include eagleeye-identityValue for /wallet/settle when available 1`] = `
{
  "basket": {
    "contents": [],
  },
  "identity": {
    "identityValue": "some-identity",
  },
  "location": {
    "incomingIdentifier": "outlet1",
    "parentIncomingIdentifier": "banner1",
  },
  "mode": "ACTIVE",
  "reference": "12345678",
}
`;

exports[`CTCartToEEBasketMapper mapOrderToWalletSettlePayload should return the payload for /wallet/settle 1`] = `
{
  "basket": {
    "contents": [],
  },
  "location": {
    "incomingIdentifier": "outlet1",
    "parentIncomingIdentifier": "banner1",
  },
  "mode": "ACTIVE",
  "reference": "12345678",
}
`;

exports[`CTCartToEEBasketMapper mapShippingMethodSkusToBasketItems should return the mapped custom basket items 1`] = `
{
  "description": "Example Shipping Discount",
  "itemUnitCost": 300,
  "itemUnitCount": 1,
  "itemUnitMetric": "EACH",
  "salesKey": "SALE",
  "totalUnitCost": 300,
  "totalUnitCostAfterDiscount": 300,
  "upc": "245879",
}
`;

exports[`CTCartToEEBasketMapper mapVoucherCodesToCampaignTokens should ignore empty string vouchers 1`] = `[]`;

exports[`CTCartToEEBasketMapper mapVoucherCodesToCampaignTokens should return an array of tokens to be examined 1`] = `
[
  {
    "type": "TOKEN",
    "value": "12345678",
  },
]
`;
